src/rules/markUsed.js

```diff
@@ -0,0 +1,98 @@
+import {
+  parse,
+} from 'jsdoc-type-pratt-parser';
+import iterateJsdoc from '../iterateJsdoc';
+
+/**
+ * Extracts the type names from parsed type declaration.
+ */
+const extractTypeNames = (parsed) => {
+  if (typeof parsed !== 'object') {
+    /* istanbul ignore next */
+    return [];
+  } else if (parsed.type === 'JsdocTypeName') {
+    return [
+      parsed.value,
+    ];
+  } else if (parsed.type === 'JsdocTypeGeneric') {
+    return [
+      ...extractTypeNames(parsed.left),
+      ...parsed.elements.flatMap(extractTypeNames),
+    ];
+  } else if (parsed.type === 'JsdocTypeFunction') {
+    return [
+      ...parsed.parameters.flatMap(extractTypeNames),
+      ...extractTypeNames(parsed.returnType),
+    ];
+  } else if (parsed.type === 'JsdocTypeNamePath') {
+    return extractTypeNames(parsed.left);
+  } else if (parsed.type === 'JsdocTypeImport') {
+    // We don't want this to fall through to the base-case
+    return [];
+  } else if (parsed.type === 'JsdocTypePredicate') {
+    // We purposefully don't consider the left (subject of the predicate) used
+    return extractTypeNames(parsed.right);
+  } else if (parsed.type === 'JsdocTypeObjectField') {
+    return [
+      ...extractTypeNames(parsed.key),
+      ...extractTypeNames(parsed.right),
+    ];
+  } else if (parsed.type === 'JsdocTypeJsdocObjectField') {
+    return [
+      ...extractTypeNames(parsed.left),
+      ...extractTypeNames(parsed.right),
+    ];
+  } else if (
+    parsed.type === 'JsdocTypeKeyValue' ||
+    parsed.type === 'JsdocTypeIndexSignature' ||
+    parsed.type === 'JsdocTypeMappedType') {
+    return extractTypeNames(parsed.right);
+  } else {
+    const result = [];
+    if (parsed.element) {
+      result.push(...extractTypeNames(parsed.element));
+    }
+
+    if (parsed.elements) {
+      result.push(...parsed.elements.flatMap(extractTypeNames));
+    }
+
+    return result;
```

Comment:
I'm using a "fall through" case here, for all the types that conformed to having a root node in `element` or a list of root nodes in `elements`.

I basically read through https://github.com/jsdoc-type-pratt-parser/jsdoc-type-pratt-parser/blob/main/src/result/RootResult.ts and https://github.com/jsdoc-type-pratt-parser/jsdoc-type-pratt-parser/blob/main/src/result/NonRootResult.ts to handle all expected cases.

Reply:
This code went away.